File: go/test/sticker_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestStickerEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Sticker(nil)
		if ent == nil {
			t.Fatal("expected non-nil StickerEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := stickerBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "sticker." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_STICKER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		stickerRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.sticker", setup.data)))
		var stickerRef01Data map[string]any
		if len(stickerRef01DataRaw) > 0 {
			stickerRef01Data = core.ToMapAny(stickerRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = stickerRef01Data

		// UPDATE
		stickerRef01Ent := client.Sticker(nil)
		stickerRef01DataUp0Up := map[string]any{
			"card_id": setup.idmap["card_id"],
		}

		stickerRef01ResdataUp0Result, err := stickerRef01Ent.Update(stickerRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		stickerRef01ResdataUp0 := core.ToMapAny(entityData(stickerRef01ResdataUp0Result))
		if stickerRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}

		// LOAD
		stickerRef01MatchDt0 := map[string]any{}
		stickerRef01DataDt0Loaded, err := stickerRef01Ent.Load(stickerRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if stickerRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func stickerBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "sticker", "StickerTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read sticker test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse sticker test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"sticker01", "sticker02", "sticker03", "card01", "card02", "card03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_STICKER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_STICKER_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_STICKER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add card_id alias for update test.
	if idmapResolved["card_id"] == nil {
		idmapResolved["card_id"] = idmapResolved["card01"]
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/sticker_entity_test.lua
-- Sticker entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("trello_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("StickerEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Sticker(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = sticker_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "sticker." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_STICKER_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local sticker_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.sticker")))
    local sticker_ref01_data = nil
    if #sticker_ref01_data_raw > 0 then
      sticker_ref01_data = helpers.to_map(sticker_ref01_data_raw[1][2])
    end

    -- UPDATE
    local sticker_ref01_ent = client:Sticker(nil)
    local sticker_ref01_data_up0_up = {
      id = sticker_ref01_data["id"],
      ["card_id"] = setup.idmap["card_id"],
    }

    local sticker_ref01_resdata_up0_result, err = sticker_ref01_ent:update(sticker_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local sticker_ref01_resdata_up0 = helpers.to_map(type(sticker_ref01_resdata_up0_result) == 'table' and sticker_ref01_resdata_up0_result.data_get and sticker_ref01_resdata_up0_result:data_get() or sticker_ref01_resdata_up0_result)
    assert.is_not_nil(sticker_ref01_resdata_up0)
    assert.are.equal(sticker_ref01_resdata_up0["id"], sticker_ref01_data_up0_up["id"])

    -- LOAD
    local sticker_ref01_match_dt0 = {
      id = sticker_ref01_data["id"],
    }
    local sticker_ref01_data_dt0_loaded, err = sticker_ref01_ent:load(sticker_ref01_match_dt0, nil)
    assert.is_nil(err)
    local sticker_ref01_data_dt0_load_result = helpers.to_map(type(sticker_ref01_data_dt0_loaded) == 'table' and sticker_ref01_data_dt0_loaded.data_get and sticker_ref01_data_dt0_loaded:data_get() or sticker_ref01_data_dt0_loaded)
    assert.is_not_nil(sticker_ref01_data_dt0_load_result)
    assert.are.equal(sticker_ref01_data_dt0_load_result["id"], sticker_ref01_data["id"])

  end)
end)

function sticker_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/sticker/StickerTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read sticker test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "sticker01", "sticker02", "sticker03", "card01", "card02", "card03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("TRELLO_TEST_STICKER_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["TRELLO_TEST_STICKER_ENTID"] = idmap,
    ["TRELLO_TEST_LIVE"] = "FALSE",
    ["TRELLO_TEST_EXPLAIN"] = "FALSE",
    ["TRELLO_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["TRELLO_TEST_STICKER_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["card_id"] == nil then
    idmap_resolved["card_id"] = idmap_resolved["card01"]
  end

  if env["TRELLO_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["TRELLO_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["TRELLO_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["TRELLO_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
